KundaNeta/assignment-3/WordSearch.java

```
@@ -0,0 +1,77 @@
+      throw new IllegalArgumentException("You must enter a defined lexicon and a defined grid.");
+    }
+
+    // notice that grid always has at least one element,
+    // and lexicon always has at least one word.
+    // this will be clarified in their respective class/constructor descriptions
+    TreeSet<String> words = new TreeSet<String>();
+    Lexicon.TrieNode lexTrieRoot = lex.getRoot();
+    int rows = grid.getRows();
+    int cols = grid.getCols();
+    int maxWords = lex.getSize();
+
+    for (int i = 0; i < rows; i++) {
+      for (int j = 0; j < cols; j++) {
+        addValidWords(grid.getCell(i, j), lexTrieRoot, lex, words);
+
+        if (words.size() == maxWords) {
+          return words;
+        }
+      }
+    }
+
+    return words;
+  }
+
+  private static void addValidWords(
+      Grid.Cell charToCheck, Lexicon.TrieNode prevCharNode, Lexicon lex, TreeSet<String> words) {
+    char currChar = charToCheck.getLetter();
+    @Nullable Lexicon.TrieNode currCharNode = prevCharNode.getChild(currChar);
+
+    // currCharNode is null if the previous string+charToCheck is not a word or a prefix
+    // in the lexicon
+    if (currCharNode == null) {
+      return;
+    }
+
+    if (currCharNode.isEndOfWord()) {
+      words.add(currCharNode.getStringSoFar());
+    }
+
+    // node.isPrefix() returns true if the node has children (if there is at least one more
+    // word in the lexicon with the current string as its prefix).
+    // also notice that a word is not a prefix of itself. this will also be clarified in the class
+    // description.
+    if (!currCharNode.isPrefix()) {
+      return;
+    }
+
+    charToCheck.isVisited = true;
+
+    for (Grid.Cell nextChar : charToCheck.getNeighboors()) {
+      if (words.size() == lex.getSize()) {
+        // break and not return to make sure we don't skip changing isVisited back to false
+        // after the loop (we could also create a resetIsValid method in Grid)
```

So this is the second place where I check if the maximum number of lexicon words has been reached (first check is in the caller method's loop). 
And I think that inside the recursive method, this is the most efficient location for the check for the purpose of preventing redundant further recursive calls.
I think that in the current implementation it isn't wasteful as is, do you think that it will be best to still look for a way to minimize the number of checks down to one?

---

There are a couple ways to do it: you could throw a custom exception, or completely reorganize the loop. :)
I think two checks are fine though.

---

Oh, an exception! Right, I hadn't thought of that! 
I think I could reorganize the loop/method so instead of a recursive method I could manually manage a stack and have the relevant calculations in a while loop of while stack is not empty OR max number of words hasn't been reached (or something similar).

But I think I'll keep it as is for now since it's okay, and will remember these options for next time!

---

By the way, one more option to do this is to make this method return a `boolean` with a meaning of "We have found everything we could", and break (or return) on that, so the condition itself would only be explicitly written in one place: exactly where you add a new word to the set.